tests/src/test/java/io/streamnative/pulsar/handlers/kop/MessagePublishBufferThrottle.java

```diff
@@ -0,0 +1,114 @@
+        Thread.sleep(20);
+        Assert.assertFalse(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
+        // Make sure the producer can publish succeed.
+        for (int i = 0; i < 10; i++) {
+            producer.send(new ProducerRecord<>(topic, new byte[1024])).get();
+        }
+        Thread.sleep(20);
+        Assert.assertFalse(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
+        super.internalCleanup();
+    }
+
+    @Test
+    public void testMessagePublishBufferThrottleEnable() throws Exception {
+        // set size for max publish buffer before broker start
+        conf.setMaxMessagePublishBufferSizeInMB(1);
+        conf.setMessagePublishBufferCheckIntervalInMillis(10000);
+        super.internalSetup();
+        Assert.assertFalse(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
+
+        final String topic = "testMessagePublishBufferThrottleEnable";
+        final String pulsarTopic = "persistent://public/default/" + topic + "-partition-0";
+        Properties properties = new Properties();
+        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:" + kafkaBrokerPort);
+        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
+        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
+        properties.put("delivery.timeout.ms", 300000);
+        KafkaProducer<byte[], byte[]> producer = new org.apache.kafka.clients.producer.KafkaProducer(properties);
+
+        // send one record to make sure InternalProducer build on broker side
+        producer.send(new ProducerRecord<>(topic, "test".getBytes())).get();
+
+        Topic topicRef = pulsar.getBrokerService().getTopicReference(pulsarTopic).get();
+        Assert.assertNotNull(topicRef);
+        InternalServerCnx internalServerCnx = (InternalServerCnx)
+                ((InternalProducer)topicRef.getProducers().values().toArray()[0]).getCnx();
+        internalServerCnx.setMessagePublishBufferSize(Long.MAX_VALUE / 2);
+        Assert.assertFalse(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
+        // The first message can publish success, but the second message should be blocked
+        producer.send(new ProducerRecord<>(topic, new byte[1024])).get(1, TimeUnit.SECONDS);
+        Thread.sleep(10000);
+        Assert.assertTrue(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
+
+        try {
+            producer.send(new ProducerRecord<>(topic, new byte[1024])).get(1, TimeUnit.SECONDS);
+            Assert.fail("Send operation should failed because of timeout.");
+        } catch (TimeoutException e) {
+            // Expected
+        }
+
+
+        internalServerCnx.setMessagePublishBufferSize(0);
+        Thread.sleep(10000);
```

Comment:
You should add comments here to indicate `sleep` is just for waiting until `checkMessagePublishBuffer` was called. Here's an example: https://github.com/apache/pulsar/blob/3142456107e8bc5d3d0d7fa526fda114c0d64c42/pulsar-broker/src/test/java/org/apache/pulsar/client/api/MessageDispatchThrottlingTest.java#L262

And use `conf.getMessagePublishBufferCheckIntervalInMillis()` here is more readable than hard coded 10000. By the way, IMO, just keep the default value (100) here is enough. You can just wait a little longer if you worried about the CI failure.

Reply:
you are right